### ISSUE TITLE
Created new dataset for user (for encryption), won't be saved

### ISSUE THREAD

**Issue by nielsk** [ubuntu/zsys#81]
I created a new dataset for a user, so that I can use zfs native encryption for the home. When I did zsysctl save it told me that there is no association for that user and did some autosave of the old filesystem. Now I destroyed the old file system and now zsysctl errors out because the file system does not exist.
But there is the encrypted file system and its mount point is the home of the user.
How do I associate a zfs with a user, so that zsysctl knows what to do?

Even more fun - and I did not try it yet: ubuntu does not create new zfs for new users but just creates new home-directories under /home which is in rpool/USERDATA thus they will not be able to save their homes. But that is a problem of ubuntu.

**Comment by didrocks** [comment]
> Even more fun - and I did not try it yet: ubuntu does not create new zfs for new users but just creates new home-directories under /home which is in rpool/USERDATA thus they will not be able to save their homes. But that is a problem of ubuntu.

Oh, this isn’t supposed to happen, Let me have a look there, adduser and gnome-control-cener and anything that is using useradd should create the new datasets. That may be a regression in focal compared to eoan. I’m having a look.

> How do I associate a zfs with a user, so that zsysctl knows what to do?
However, indeed, if you want special options like per dataset encryption, this isn’t supported directly with zsysctl yet for focal. You can associate it quite easily though:
1. get your current root dataset path: (with zfs mount for instance), let’s say it’s `rpool/ROOT/ubuntu_12345678`
2. Tag your newly created user dataset (which should start with username, for instance `rpool/USERDATA/user1_abcdef` to associate with the root pool: `zfs set com.ubuntu.zsys:bootfs-datasets=rpool/ROOT/ubuntu_12345678 rpool/ROOT/ubuntu_12345678`
3. That should be it, if you want to check that everything is alright:
- if no zsysd is running, just run `zsysctl show --full` to list user datasets
- if zsysd was still running (it’s exiting after one minute of idling), `zsysctl service refresh && zsysctl show --full`.
We should make an utility for this until we have a better user-experience for it.

Keep us posted!

**Comment by didrocks** [comment]
> > Even more fun - and I did not try it yet: ubuntu does not create new zfs for new users but just creates new home-directories under /home which is in rpool/USERDATA thus they will not be able to save their homes. But that is a problem of ubuntu.
> 
> Oh, this isn’t supposed to happen, Let me have a look there, adduser and gnome-control-cener and anything that is using useradd should create the new datasets. That may be a regression in focal compared to eoan. I’m having a look.

You’re right, I pushed a fix which will be building in focal shortly (shadow package 4.8.1-1ubuntu4) once accepted by the freeze.

**Comment by nielsk** [comment]
Thanks. I will try that out later. I just broke an installation and have to re-install because I played around too much.

**Comment by didrocks** [comment]
Please keep breakikng it and report anything that is weird! :)

**Comment by nielsk** [comment]
> Tag your newly created user dataset (which should start with username, for instance rpool/USERDATA/user1_abcdef to associate with the root pool: zfs set com.ubuntu.zsys:bootfs-datasets=rpool/ROOT/ubuntu_12345678 rpool/ROOT/ubuntu_12345678

I guess this is 
```
zfs set com.ubuntu.zsys:bootfs-datasets=rpool/ROOT/ubuntu_12345678 rpool/USERDATA/user1_abcdef
```

> We should make an utility for this until we have a better user-experience for it.

Or mention it in the man-page of zsys. I didn't look up the zfs-properties because I totally didn't expect it there.

**Comment by nielsk** [comment]
meh. There are no real man-pages. And it seems you can only find out about that property with a 
```
zfs get all
```

**Comment by nielsk** [comment]
This seems to work.
I had to destroy the old dataset and do another refresh to get no errors anymore.

**Comment by didrocks** [comment]
Oh right, sorry about the typo :) And indeed, you need to deassociate curent dataset.
There will be a series of blog posts (more and less technical) once 20.04 LTS is out to cover those aspects.

The shadow package is uploaded which will fix "create userdataset on demand", but manual ones like you did won't until we have a better way to enroll user dataset. I’ll retitle this bug for this.
Keep bug reports coming if you spot anything else bad!

**Comment by nielsk** [comment]
Blog posts are great but having manuals on my system without the need to google are far better because I do not have to rely on search engines but use the tools available on my system.

**Comment by didrocks** [comment]
It’s hard as implementer to know what is really technical details or what could be interesting to advanced users who wants to understand a little bit more and to some manual work, while still being compatible with the design.

If I write those blog posts would you be interested in helping crafting those manuals (probably in markdown) here so that we can ship them as man pages in the distro?

**Comment by nielsk** [comment]
I might try to help out. I am not a technical writer but I can give it a a shot.

**Comment by FliegenKLATSCH** [comment]
I first tried `zsysctl userdata set-home` but finally found this issue...
Some `zsysctl userdata xxx` alias would be great for this simple task.
Maybe `set-home` should be named `rename` and the `set-home` task should set this property?

**Comment by didrocks** [comment]
Those are hidden commands on purpose: just to change the user home directory, please use rather `usermod -m -d` which is the command intended to it (it refreshes AccountServices and other things) to change the user home directory.

The command is hooked up to `zsysctl userdata set-home` which DTRT.

**Comment by FliegenKLATSCH** [comment]
I had the same use-case as the issue creator: I manually created a new encrypted home volume and renamed it with `zfs` to match the old name. I just noticed on apt update's a warning and investigated... I think encrypted home folders will become a popular feature because with zfs it's quite easy to archive..

**Comment by didrocks** [comment]
Agreed! But the fix is thus different: set-home is just a workaround, we need a dedicated command to enroll existing home directory (if created under `USERDATA`) filesystem datasets.

I think something like `zsysctl machine add-user-dataset` or something like that? (This command would check if it’s under `USERDATA`, and then just tag it to associate to current system + unassociate (after a prompt, potenially) any other dataset associated for this user with this machine.
None of this check is done with `set-home` as its purpose is different and some checks are already done beforehand by the usermod command.

On encryption, agreed, and this is why we are working on bringing to 20.10 encryption as an easy option by default! (This will be probably backported to 20.04.x release).

**Comment by mcamou** [comment]
I have a similar problem. I installed Ubuntu 19.10 from scratch and set it up with ZFS. Since I wanted to encrypt my home directory (and possibly others) I set up an `rpool/encrypted` dataset and moved my home directory there. I also have other datasets under my home directory so that I can snapshot certain parts of it.

I also moved `rpool/ROOT/var/lib/docker` to `rpool/docker` just to keep the output of `zfs list` more manageable, since I am constantly running different Docker containers and `zfs list` ends up really cluttered.

I have now updated to 20.04 and find that the choices I made with 19.04 result in `zsys` not picking up my home directory and `/var/lib/docker`. I would really like to add directories under `rpool/encrypted` and `rpool/docker` to zsys.
